app/lib/service/email/backend.dart

```diff
@@ -62,84 +60,81 @@ class EmailBackend {
   /// are saved in the Datastore.
-  Future<void> trySendOutgoingEmails(Iterable<OutgoingEmail> emails) async {
-    for (final id in emails.map((e) => e.uuid)) {
-      await _trySendOutgoingEmail(id);
-    }
+  Future<void> trySendOutgoingEmail(OutgoingEmail email) async {
+    await _trySendOutgoingEmail(email.uuid);
   }
 
   /// Tries to send email with the given [id]. The
   /// [OutgoingEmail] entry will be deleted after the
   /// email was sent successfully.
   ///
-  /// Returns true if the email was sent successfully.
-  Future<bool> _trySendOutgoingEmail(String id) async {
+  /// Returns the number of emails that were sent successfully.
+  Future<int> _trySendOutgoingEmail(String id) async {
     final key = _db.emptyKey.append(OutgoingEmail, id: id);
     final now = clock.now().toUtc();
+    final claimId = createUuid();
     final entry = await withRetryTransaction(_db, (tx) async {
       final o = await tx.lookupOrNull<OutgoingEmail>(key);
-      if (o == null || o.isNotAlive) {
+      if (o == null || o.isNotAlive || o.claimId != null) {
         return null;
       }
       o.attempts++;
       o.lastAttempted = now;
       // retry after a random delay in the next 2-6 hours
       o.pendingAt =
           now.add(Duration(hours: 2, minutes: _random.nextInt(4 * 60)));
+      o.claimId = claimId;
       tx.insert(o);
-
       return o;
     });
     if (entry == null) {
-      return false;
+      return 0;
     }
 
-    try {
-      await emailSender.sendMessage(EmailMessage(
-        uuid: entry.uuid,
-        EmailAddress(entry.fromEmail!),
-        [EmailAddress(entry.recipientEmail!)],
-        entry.subject!,
-        entry.bodyText!,
-      ));
-
-      await withRetryTransaction(_db, (tx) async {
-        final o = await tx.lookupOrNull<OutgoingEmail>(key);
-        if (o != null) {
-          tx.delete(key);
-        }
-      });
+    final recipientEmails = entry.recipientEmails ?? const <String>[];
+    final sent = <String>[];
+    for (final recipientEmail in recipientEmails) {
+      try {
+        await emailSender.sendMessage(EmailMessage(
+          uuid: entry.uuid,
+          EmailAddress(entry.fromEmail!),
+          [EmailAddress(recipientEmail)],
+          entry.subject!,
+          entry.bodyText!,
+        ));
+        sent.add(recipientEmail);
+      } catch (e, st) {
+        _logger.warning('Email sending failed (claimId="$claimId").', e, st);
+      }
+    }
 
-      // successful send, deleting entry
-      return true;
-    } catch (e, st) {
-      _logger.warning('Email sending failed.', e, st);
-      await withRetryTransaction(_db, (tx) async {
-        final o = await tx.lookupOrNull<OutgoingEmail>(key);
-        if (o == null) {
-          return;
-        }
-        o.lastError = e.toString();
+    await withRetryTransaction(_db, (tx) async {
+      final o = await tx.lookupOrNull<OutgoingEmail>(key);
+      if (o == null || o.claimId != claimId) {
+        return;
```

Comment:
This needs to log something with level shout or something... because this is really bad!